src/bin/nydus-image/merge.rs

```diff
@@ -115,26 +125,35 @@ impl Merger {
                 flags = Some(current_flags);
             }
 
-            let parent_blobs = rs.superblock.get_blob_infos();
             let blob_hash = Self::get_blob_hash(bootstrap_path)?;
             let mut blob_idx_map = Vec::new();
             let mut parent_blob_added = false;
-
-            for blob in &parent_blobs {
-                let mut blob_ctx = BlobContext::from(ctx, blob, ChunkSource::Parent);
+            for blob in rs.superblock.get_blob_infos() {
+                let mut blob_ctx = BlobContext::from(ctx, &blob, ChunkSource::Parent);
                 if chunk_dict_blobs.get(blob.blob_id()).is_none() {
                     // It is assumed that the `nydus-image create` at each layer and `nydus-image merge` commands
                     // use the same chunk dict bootstrap. So the parent bootstrap includes multiple blobs, but
                     // only at most one new blob, the other blobs should be from the chunk dict image.
                     if parent_blob_added {
-                        bail!("invalid bootstrap, seems have multiple non-chunk-dict blobs in this bootstrap");
+                        bail!("invalid per layer bootstrap, having multiple associated data blobs");
                     }
                     // The blob id (blob sha256 hash) in parent bootstrap is invalid for nydusd
                     // runtime, should change it to the hash of whole tar blob.
                     blob_ctx.blob_id = blob_hash.to_owned();
-                    chunk_size = Some(blob_ctx.chunk_size);
+                    if chunk_size.is_none() {
+                        chunk_size = Some(blob_ctx.chunk_size);
+                    }
+                    if chunk_size != Some(blob_ctx.chunk_size) {
+                        bail!(
+                            "can not merge bootstraps with inconsistent chunk size, current bootstrap {:?} with chunk size {:x}, expected {:x}",
+                            bootstrap_path,
+                            blob_ctx.chunk_size,
+                            chunk_size.unwrap(),
+                        );
+                    }
                     parent_blob_added = true;
                 }
+                // TODO: this seems wrong, blobs from chunk dict will be added more than one time.
```

Comment:
What's the problem here?

Reply:
Fixed

Reply:
Say we are going to merge two images, and those two image reference a common layer (LayerA), then there will be two LayerA in the merged blob table.

Reply:
We need something like this
```
diff --git a/src/bin/nydus-image/merge.rs b/src/bin/nydus-image/merge.rs
index 3b25398f2..bf63ebd7d 100644
--- a/src/bin/nydus-image/merge.rs
+++ b/src/bin/nydus-image/merge.rs
@@ -153,9 +153,18 @@ impl Merger {
                     }
                     parent_blob_added = true;
                 }
-                // TODO: this seems wrong, blobs from chunk dict will be added more than one time.
-                blob_idx_map.push(blob_mgr.len() as u32);
-                blob_mgr.add(blob_ctx);
+
+                let mut found = false;
+                for (idx, blob) in blob_mgr.get_blobs().iter().enumerate() {
+                    if blob.blob_id == blob_ctx.blob_id {
+                        blob_idx_map.push(idx as u32);
+                        found = true;
+                    }
+                }
+                if !found {
+                    blob_idx_map.push(blob_mgr.len() as u32);
+                    blob_mgr.add(blob_ctx);
+                }
             }
 
             if let Some(tree) = &mut tree {
```

Reply:
Okay, good catch!